docker/daemon/client_test.go

```diff
@@ -0,0 +1,175 @@
+fgZBjocNt8pJdA/KaqGmiRxVzayAqRIME673nWCRcKp9y6Ih3Bd2sjbMtuavtp2C
+YBZF1xxBgNZQaZ8WJxPnnQECgYEAzLgGJPWc5iyZCJsesQTbMICRTyEPTYKKFD6N
+6CFt+vDgNsUxOWRx0Vk6kUhW+rAItZzjgZ6RBzyuwtH17sGYZHZefMZL4Y2/QSru
+ej/IpNRjwaF6AN0KxhfhXcCw8zrivX/+WgqOcJj7lh/TC7a/S0uNNSgJ5DODKwd9
+WSboPvkCgYEAzqdWfetko7hEI4076pufJrHPnnCJSHkkQ1QnfVl71mq7UmKXLDxD
+L5oWtU53+dswzvxGrzkOWsRJC5nN30BYJuYlwKzo3+MCKlUzJSuIMVTbTPlwKudh
+AF19s4GFZVo29FlgIQhA5dfIkZgFXAlVxYcGTLUixEmPwrc6yguULPs=
+-----END RSA PRIVATE KEY-----
+`
+
+func TestDockerClientFromNilSystemContext(t *testing.T) {
+	client, err := newDockerClient(nil)
+
+	assert.Nil(t, err, "There should be no error creating the Docker client")
+	assert.NotNil(t, client, "A Docker client reference should have been returned")
+
+	assert.Equal(t, dockerclient.DefaultDockerHost, client.DaemonHost(), "The default docker host should have been used")
+	assert.Equal(t, api.DefaultVersion, client.ClientVersion(), "The default api version should have been used")
+}
+
+func TestDockerClientFromCertContext(t *testing.T) {
+	tmpDir, err := ioutil.TempDir("", "image-test-docker-client")
+	if err != nil {
+		t.Fatal("Unable to create tmp directory")
+	}
+	defer os.RemoveAll(tmpDir)
+	certSetup(t, tmpDir)
+
+	host := "tcp://127.0.0.1:2376"
+	version := "1.11"
+	systemCtx := &types.SystemContext{
+		DockerCertPath:   tmpDir,
+		DockerHost:       host,
+		DockerAPIVersion: version,
+		DockerTLSVerify:  true,
+	}
+
+	client, err := newDockerClient(systemCtx)
+
+	assert.Nil(t, err, "There should be no error creating the Docker client")
+	assert.NotNil(t, client, "A Docker client reference should have been returned")
+
+	assert.Equal(t, host, client.DaemonHost())
+	assert.Equal(t, version, client.ClientVersion())
+}
+
+func TestTlsConfigFromNilSystemContext(t *testing.T) {
+	httpClient, err := tlsConfig(nil)
+
+	assert.Nil(t, err, "There should be no error creating the HTTP client")
+	assert.Nil(t, httpClient, "With no explicit TLS configutation, there should be no HTTP client")
+}
+
+func TestTlsConfigFromInvalidCertPath(t *testing.T) {
+	systemCtx := &types.SystemContext{
+		DockerCertPath: "/foo/bar",
+	}
+
+	_, err := tlsConfig(systemCtx)
+
+	if assert.Error(t, err, "An error was expected") {
+		assert.Regexp(t, "could not read CA certificate", err.Error())
+	}
+}
+
+func TestTlsConfigFromCertPath(t *testing.T) {
+	tmpDir, err := ioutil.TempDir("", "image-test-docker-client")
+	if err != nil {
+		t.Fatal("Unable to create tmp directory")
+	}
+	defer os.RemoveAll(tmpDir)
+	certSetup(t, tmpDir)
+
+	systemCtx := &types.SystemContext{
+		DockerCertPath:  tmpDir,
+		DockerTLSVerify: true,
+	}
+
+	httpClient, err := tlsConfig(systemCtx)
+
+	assert.Nil(t, err, "There should be no error creating the HTTP client")
+
+	transport := httpClient.Transport.(*http.Transport)
+	assert.True(t, transport.TLSClientConfig.InsecureSkipVerify, "TLS verification should be skipped")
+}
+
+func certSetup(t *testing.T, tmpDir string) {
```

Comment:
Instead, create a `fixtures/certs` subdirectory inside the package, and refer to it by path. No need to create temporary directories, create copies of the files, or to carry them as strings inside Go code.

Reply:
I guess its a matter of choice, no?

Reply:
It’s much more code with embedding everything into `*.go`, and it’s more difficult to inspect the certificates. Are there any upsides to doing it this way?